### PR TITLE
Improved emailvalidator & tests

### PR DESCRIPTION
**Pull request by SamMousa**
This is a first commit in fixing the emailvalidator.
Instead of having several different tests that test different sets of email addresses, I've refactored the code to test 1 list of email addresses more thoroughly.

The reasoning is this:
1. An email address like: `Bob <bob@example.com>` is either valid or not.
2. It might require IDN to be validated.
3. It might contain a name to be validated.

Each email is then tested against all scenarios:
1. If we claim it requires IDN, it MUST NOT validate when we haven't enabled IDN.
2. If we claim it contains a name it MUST NOT validate when names are not allowed.

The tests currently fail for several reasons:
1. I've added an error case that identifies the original issue causing this PR (expected).
2. It seems some email addresses are not properly validate while they should be.

Specifically:
`example@xn--zcack7ayc9a.de`, as far as I can tell, this is a valid email address that does not require IDN and does not contain a name. 
`после-преобразования-в-idn-тут-будет-больше-чем-64-символа@пример.com`, as far as I can tell, is a valid email address that requires IDN and does not have a name.

I have validated both email addresses via this tool https://verifalia.com/validate-email, which agrees that both email addresses are syntactically valid.
Note that the second one is valid, but then fails due to extended, ISP specific, validation. 

| Q             | A
| ------------- | ---
| Is bugfix?    | yes
| New feature?  | no
| Breaks BC?    | no
| Tests pass?   | no
| Fixed issues  | #15947

### CODE REVIEW

**Comment by samdark** [comment]
It fails.

**Comment by SamMousa** [comment]
Yes, it is expected fail, it's not finished yet, but i had to remove WIP to get it to run the tests..

File: tests/framework/validators/EmailValidatorTest.php [hunk 1]
```diff
@@ -24,87 +24,89 @@ protected function setUp()
...
+            [true, false, false, 'sam@rmcreative.ru'],
+            [true, false, false, 'test-@example.com'],
+            [true, false, false, 'test@example.com'],
+            [true, false, false, 'user+mailbox/department=shipping@example.com'],
+            [true, false, true, '"Carsten Brandt" <mail@cebe.cc>'],
+            [true, false, true, '"This name is longer than 64 characters. Blah blah blah blah blah" <shortmail@example.com>'],
+            [true, false, true, '"Такое имя достаточно длинное, но оно все равно может пройти валидацию" <shortmail@example.com>'],
+            [true, false, true, '<mail@cebe.cc>'],
+            [true, false, true, 'Carsten Brandt <mail@cebe.cc>'],
+            [true, false, true, 'example@xn--zcack7ayc9a.de'],
```
Comment (samdark) [inline tests/framework/validators/EmailValidatorTest.php]: The assertion is correct. It's a valid email.

File: tests/framework/validators/EmailValidatorTest.php [hunk 2]
```diff
@@ -24,87 +24,89 @@ protected function setUp()
...
+            [true, false, true, 'John Smith <john.smith@example.com>'],
+            [true, true, false, 'example@äüößìà.de'],
+            [true, true, false, 'info@örtliches.de'],
+            [true, true, false, 'sam@рмкреатиф.ru'],
+            [true, true, false, 'üñîçøðé@üñîçøðé.com'],
+            [true, true, true, 'Information <info@örtliches.de>'],
+            [true, true, true, 'üñîçøðé 日本国 <üñîçøðé@üñîçøðé.com>'],
+            [true, true, false, 'после-преобразования-в-idn-тут-будет-больше-чем-64-символа@пример.com'],
```
Comment (samdark) [inline tests/framework/validators/EmailValidatorTest.php]: This is not valid email. Local part is too long after encoding into IDN.

File: tests/framework/validators/EmailValidatorTest.php [hunk 3]
```diff
@@ -24,87 +24,89 @@ protected function setUp()
...
+            [true, true, true, 'Короткое имя <после-преобразования-в-idn-тут-будет-больше-чем-64-символа@пример.com>'],
```
Comment (samdark) [inline tests/framework/validators/EmailValidatorTest.php]: This is not valid email. Local part is too long after encoding to IDN.

**Comment by SamMousa** [comment]
@samdark I've changed the flow of the code in the validator.
Hmm; apparently that's exactly what we don't want :D 
I'll change that back, could you please check if my regex for splitting the email is valid? 
And if it is not valid, could you provide some email addresses so I can add them as test cases?

**Comment by samdark** [comment]
- https://stackoverflow.com/questions/7641019/list-of-mock-valid-invalid-email-addresses-for-unit-tests

**Comment by samdark** [comment]
http://criticaltester.com/test-data/email-address/

**Comment by SamMousa** [comment]
> The changes you've made make it impossible to merge it into 2.0 at all and also I don't like the increased configuration complexity.

I just went with configuration because the original patterns use configuration. In my opinion these should be constants or variables defined inside the validation function.
The reasoning for splitting them was actually to make it more readable not necessarily configurable.

**Comment by samdark** [comment]
The initial idea of having a single regexp was an ability to make it configurable. You may have more restrictive emails policy. For example, accept emails only from certain domains.

**Comment by samdark** [comment]
Doesn't seem to be merge-able. @SamMousa do you want to finish it or close it?

**Pull request closed without merging**